src/net_processing.h

```diff
@@ -89,7 +103,314 @@ struct Peer {
+                    } else {
+                        local_set_size = m_local_set.size();
+                    }
+                    uint8_t remote_set_size = local_set_size + actual_local_missing - actual_remote_missing;
+                    uint8_t set_size_diff = std::abs(local_set_size - remote_set_size);
+                    uint8_t min_size = std::min(local_set_size, remote_set_size);
+                    uint8_t actual_difference = actual_local_missing + actual_remote_missing;
+                    if (min_size != 0) {
+                        m_local_q = double(actual_difference - set_size_diff) / min_size;
+                        assert(m_local_q >= 0 && m_local_q <= 2);
+                    }
+                } else if (action == LocalQAction::Q_SET_DEFAULT) {
+                    m_local_q = DEFAULT_RECON_Q;
+                }
+                m_outgoing_recon = ReconPhase::NONE;
+            }
+            if (clear_local_set) m_local_set.clear();
+
+            m_local_short_id_mapping.clear();
+            // This is currently belt-and-suspenders, as the code should work even without these calls.
+            m_remote_sketch_snapshot.clear();
+            m_local_set_snapshot.clear();
+            m_capacity_snapshot = 0;
+        }
+
+        /**
+         * When during reconciliation we find a set difference successfully (by combining sketches),
+         * we want to find which transactions are missing on our and on their side.
+         * For those missing on our side, we may only find short IDs.
+         */
+        std::vector<uint256> GetRelevantIDsFromShortIDs(std::vector<uint64_t> diff, std::vector<uint32_t>& local_missing)
+        {
+            std::vector<uint256> remote_missing;
+            for (const auto& diff_short_id: diff) {
+                const auto local_tx = m_local_short_id_mapping.find(diff_short_id);
+                if (local_tx != m_local_short_id_mapping.end()) {
+                    remote_missing.push_back(local_tx->second);
+                } else {
+                    local_missing.push_back(diff_short_id);
+                }
+            }
+            return remote_missing;
+        }
+
+        /**
+         * After a reconciliation round passed, transactions missing by our peer are known by short ID.
+         * Look up their full wtxid locally to announce them to the peer.
+         */
+        std::vector<uint256> GetWTXIDsFromShortIDs(const std::vector<uint32_t>& remote_missing_short_ids)
+        {
+            std::vector<uint256> remote_missing;
+            for (size_t i = 0; i < remote_missing_short_ids.size(); ++i) {
+                auto local_tx = m_local_short_id_mapping.find(remote_missing_short_ids[i]);
+                if (local_tx != m_local_short_id_mapping.end()) {
+                    remote_missing.push_back(local_tx->second);
+                }
+            }
+            return remote_missing;
+        }
+    };
+
+    RecursiveMutex cs_recon_state;
```

Comment:
Don't use old cs nomenclature for mutexes:

```suggestion
    RecursiveMutex m_recon_state_mutex;
```

Also prefer to use a `Mutex` over a `RecursiveMutex` in new code.

Reply:
I think I can't make it non-recursive because of the call inside `GetFloodingOutboundsCount`. How would you suggest to overcome this issue?